modules/servers/upCloudVps/lib/upCloudVps.php

```diff
@@ -234,6 +235,35 @@ public function CreateServer($ZoneID, $Hostname, $Plan, $OsUUID, $sshKey, $user_
   ];
 }
 
+if ($TemplateType == 'native') {
+    if (!preg_match('/Windows/', $TemplateTitle)) {
+        if ($sshKey != "na") {
+            $postData['server']['login_user'] = [
+                'username' => 'root',
+                'ssh_keys' => [
+                    'ssh_key' => [
+                        $sshKey,
+                    ],
+                ],
+            ];
+        }
+    }
+} elseif ($TemplateType == 'cloud-init') {
+    if ($sshKey != "na") {
+        $postData['server']['login_user'] = [
+            'username' => 'root',
+            'ssh_keys' => [
+                'ssh_key' => [
+                    $sshKey,
+                ],
+            ],
+        ];
+    } else {
+        $error['response']['error']['error_message'] = 'ssh key required';

```

Comment:
This check happens too late from good UX point of view -- the ssh key field is not marked as required in the UI for clients, this just makes the creation fail later without the client getting notified about it.

Reply:
A workaround for now is to use the admin UI to configure the ssh key as required in products using the Linux w/ cloud-init template, https://docs.whmcs.com/clients/the-client-area/custom-fields/#add-a-custom-field

Reply:
IMHO good enough for the first release - each plan requires a product so can be configured while at it 👍